File: algorithms/python/russianDollEnvelopes/russianDollEnvelopes.py
#! /usr/bin/env python3
# -*- coding: utf-8 -*-

# Source: https://leetcode.com/problems/russian-doll-envelopes/
# Author: Miao Zhang
# Date:   2021-02-04

from bisect import bisect_left
class Solution:
    def maxEnvelopes(self, envelopes: List[List[int]]) -> int:
        envelopes.sort(key=lambda x: (x[0], -x[1]))
        dp = []
        for en in envelopes:
            idx = bisect_left(dp, en[1])
            if idx == len(dp):
                dp.append(en[1])
            else:
                dp[idx] = en[1]
        return len(dp)

File: algorithms/cpp/russianDollEnvelopes/russianDollEnvelopes.cpp
// Source: https://leetcode.com/problems/russian-doll-envelopes/
// Author: Miao Zhang
// Date:   2021-02-04

class Solution {
public:
    int maxEnvelopes(vector<vector<int>>& envelopes) {
        vector<int> dp;
        sort(envelopes.begin(), envelopes.end(), [](vector<int> &a, vector<int> &b) {
            if (a[0] == b[0]) return a[1] > b[1];
            return a[0] < b[0];
        });
        for (int i = 0; i < envelopes.size(); i++) {
            auto it = lower_bound(dp.begin(), dp.end(), envelopes[i][1]);
            if (it == dp.end()) dp.push_back(envelopes[i][1]);
            else *it = envelopes[i][1];
        }
        return dp.size();
    }
};
